flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericAtLeastOnceSink.java

```diff
@@ -0,0 +1,192 @@
+	private final CheckpointCommitter committer;
+	private transient AbstractStateBackend.CheckpointStateOutputView out;
+	protected final TypeSerializer<IN> serializer;
+	private final String id;
+
+	private ExactlyOnceState state = new ExactlyOnceState();
+
+	public GenericAtLeastOnceSink(CheckpointCommitter committer, TypeSerializer<IN> serializer, String jobID) throws Exception {
+		this.committer = committer;
+		this.serializer = serializer;
+		this.id = UUID.randomUUID().toString();
+		this.committer.setJobId(jobID);
+		this.committer.createResource();
+	}
+
+	@Override
+	public void open() throws Exception {
+		committer.setOperatorId(id);
+		committer.setOperatorSubtaskId(getRuntimeContext().getIndexOfThisSubtask());
+		committer.open();
+	}
+
+	public void close() throws Exception {
+		committer.close();
+	}
+
+	/**
+	 * Saves a handle in the state.
+	 *
+	 * @param checkpointId
+	 * @throws IOException
+	 */
+	private void saveHandleInState(final long checkpointId, final long timestamp) throws Exception {
+		//only add handle if a new OperatorState was created since the last snapshot
+		if (out != null) {
+			StateHandle<DataInputView> handle = out.closeAndGetHandle();
+			if (state.pendingHandles.containsKey(checkpointId)) {
+				//we already have a checkpoint stored for that ID that may have been partially written,
+				//so we discard this "alternate version" and use the stored checkpoint
+				handle.discardState();
+			} else {
+				state.pendingHandles.put(checkpointId, new Tuple2<>(timestamp, handle));
+			}
+			out = null;
+		}
+	}
+
+	@Override
+	public StreamTaskState snapshotOperatorState(final long checkpointId, final long timestamp) throws Exception {
+		StreamTaskState taskState = super.snapshotOperatorState(checkpointId, timestamp);
+		saveHandleInState(checkpointId, timestamp);
+		taskState.setFunctionState(state);
+		return taskState;
+	}
+
+	@Override
+	public void restoreState(StreamTaskState state, long recoveryTimestamp) throws Exception {
+		super.restoreState(state, recoveryTimestamp);
+		this.state = (ExactlyOnceState) state.getFunctionState();
+
+		out = null;
+	}
+
+	@Override
+	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
+		super.notifyOfCompletedCheckpoint(checkpointId);
+
+		synchronized (state.pendingHandles) {
+			Set<Long> pastCheckpointIds = state.pendingHandles.keySet();
+			Set<Long> checkpointsToRemove = new HashSet<>();
+			for (Long pastCheckpointId : pastCheckpointIds) {
+				if (pastCheckpointId <= checkpointId) {
+					if (!committer.isCheckpointCommitted(pastCheckpointId)) {
+						Tuple2<Long, StateHandle<DataInputView>> handle = state.pendingHandles.get(pastCheckpointId);
+						DataInputView in = handle.f1.getState(getUserCodeClassloader());
+						sendValues(new ReusingMutableToRegularIteratorWrapper<>(new InputViewIterator<>(in, serializer), serializer), handle.f0);
+						committer.commitCheckpoint(pastCheckpointId);
```

Comment:
Can it be that committing the checkpoint id can fail (with cassandra)? Is the commit checkpoint operation synchronous (CassandraCommitter)?